### ISSUE TITLE
[Notifications] Add notification infrastructure

### ISSUE THREAD

**Issue by VWoeltjen** [nasa/openmctweb#155]
Implement notification infrastructure which allows notifications to be issued to the user, per UI documents.

This should include both:

* Blocking dialogs which provide error or warning messages which the user must dismiss and acknowledge.
* Non-blocking notifications that appear in/near the status area.

Suggestion would be to implement this as one or more services (e.g. a `notificationService` with a `notify` method which takes a string `message` as a parameter and some `options` object to specify things like blocking, severity level, etc - review UI documents to determine necessary properties here.)

Thinking this should add new extensions to `platform/commonUI/dialog`, and/or to a new `platform/commonUI/notification` bundle.

**Comment by VWoeltjen** [comment]
Plan to assign to @akhenry once his user account is active

**Comment by VWoeltjen** [comment]
Regarding user interface for this, keep an eye on #154 which may add additional detail to this. (If #154 is still pending by the time this issue gets attention, though, we can handle those additional details as follow up issues.)

**Comment by larkin** [comment]
@akhenry Assigning to you.  I took some notes from the meeting this morning, take them as you will or throw them out:

# Notification Service

potential interface: has one method, "notify", takes notify object.  potential options for notify object:

 * message: content of notification in status bar, header of message in list view. 
 * details: a more details explaination to show in list view.
 * autoDismiss: boolean or integer.  if true, dismiss after default interval.  if integer, dismiss after specified number of milliseconds.  if false, notification cannot be dismissed.
 * type (or severity), 'alert', 'progress', 'green happy thing', etc.
 * progress: integer, progress out of 100, if undefined, no progress bar is shown.
 * primaryAction: if defined, show a button in the status bar.  actual value is an object with a text label and a function to call when the button is pressed.
 * customActions: array of objects, each representing an action button to show in alert overview.

# Notification Minimizing

Minimizing notifications to smaller icons to display many of them.  May be necessary to free up real estate or handle multiple notifications.  For review purposes and keeping features small, may deliver as an iteration on top of the notification service.

**Comment by larkin** [comment]
As far as blocking dialogues, I would recommend implementing those separately from notifications.  I see them as having use outside of notifications.  Will be sticking my notes for those in a separate issue.

**Comment by akhenry** [comment]
Blocking dialogues could be implemented as a separate bundle that the notifications framework then depends upon. I agree, there are other potential use cases for a generic modal dialog.

Agree with the rough design provided above. A few points
- Not sure if 'progress' should be a type / severity or just a special case of a more generic 'info' dialog? 
- Suspect we also need a 'progessMessage' attribute on the model that can be updated from time to time (eg. '10 of 15 objects duplicated'). This could alternatively be used by client code to update time estimate, time remaining message.
- presume Angular data binding will take care of updates in the UI when the value of the progress attribute is updated.

**Comment by larkin** [comment]
Think you're conflating "progress dialog" and "a notification with progress", I would choose to separate them:
* `progressDialog` is a `template` extension that supports a few specific attributes on it's model-- related to.  This can be displayed with `overlayService.createOverlay('progressDialog', progressModel);`.  (see #163)
* `a notification with progress`: any notification (of any severity) could have progress.  we may not have a use case, but we certainly don't need to spend time restricting progress to a specific severity.

In the designs for notifications, I don't see a place where progressMessage could appear, but it could be easily supported.  I suggested `progressText` in #163, but I'm fine with either-- as long as it's consistent.

**Comment by akhenry** [comment]
Agreed, all of those comments were intended for the blocking dialog. Missed the separate issue.

**Comment by akhenry** [comment]
I have pushed an initial testing version of the notifications and dialogs framework (see branch [open155](https://github.com/nasa/openmctweb/tree/open155) ). The push includes a 'testing' bundle that adds some indicators in the bottom of the application for launching dialogs and notifications of different types. This will of course be removed prior to merge.

No existing process have been modified to use the new notifications framework, or the new modal dialog type.

Blocking dialogs are invoked by calling the `showBlockingMessage` function on `DialogService`. Banner notifications are invoked by using the `notify` function on the new `NotificationService`. Note that there is also a convenience function 'info' that calls `notify` with some defaults (such as auto-minimize) appropriate for purely informational messages.

There are a couple of outstanding issues that I haven't addressed yet, and might need some discussion -
* Currently all messages are minimized to the notifications list instead of dismissed, including 'info' messages. It is probably desirable not to minimize 'info' level messages, but rather just dismiss them. This might require a little thought though, because you may want success type messages to be dismissible, but progress messages to be dismissible only once the underlying process has completed. You also probably don't want success messages cluttering up your notifications list. Or maybe you do, but you provide the ability to filter the displayed messages. Incidentally, 'progress' and 'severity' are considered separate properties, and you may have a progress message of any severity.
* The framework does not yet support clicking text within messages. The simplest way of supporting this is to allow html to be included in message text. This potentially enables an exploit if the source of the message is not carefully considered by developers using it, so I'd like to leave it as an open issue for now until we can have some discussion about the best approach here.

**Comment by larkin** [comment]
@akhenry best way to get feedback on those would be to open a pull request, so we can discuss within context of code.

As far as your testing bundle; that sounds very useful to keep around.  I would put that in the `example` folder so that we can use it in the future.

**Comment by akhenry** [comment]
This is closed by merged pull request https://github.com/nasa/openmctweb/pull/186

**Comment by RCarvalhoNASA** [comment]
Testathon 20151123:  Notification comes up and then transitions to progress bar, though I'm not sure why it does it both ways and whether one allows a different state than the other.
